### ISSUE TITLE
How can I package Douane into a .deb installer once I've finished compiling it?

### ISSUE THREAD

**Issue by thebunnyrules** [Douane/Douane#95]
Hi ZedTux,

I noticed that you were kind enough to include the files needed for making .deb installers for the different douane packages once we're done building them but I am unsure how to go about it.

**Comment by zedtux** [comment]
Hey @thebunnyrules,

Well, at the very beginning I was taking care of packaging Douane for Debian/Ubuntu, but it was really time consuming so I decided to stop doing it, and focus on the development.

I'm looking for packagers ! As you can see from [the Packages Wiki page](https://github.com/Douane/Douane/wiki/Packages), I'm missing packagers a lot 😞.

If you wants to take care of the Debian / Ubuntu packaging, I would be more than happy to have your name here 😃.

I can help you with all I remember if you need.

Tell me how much you know packing apps and I'll guide you as much as I can 👍.

**Comment by thebunnyrules** [comment]
I've never packaged before but I'd really like to learn how. It would be alot more practical than having to compile from source on every machine I want to install. If you can show me how to do it, I wouldn't mind helping out with the packaging (or writing a more detailed tutorial for the wiki). 

I found this simple tutorial here: https://ubuntuforums.org/showthread.php?t=910717 but packaging for the big distros is apparently more complicated. Speaking of which, aren't they supposed to compile and  package everything themselves? 

Anyway, from the tutorial: it seems I just need to make a folder and fill it as if were the root folder (so make an ./opt/douane/... folder and so on). Put the debian folder I found on the github page and then do:

dpkg-deb --build <name_of_project>

The thing is, how do I figure out what file goes where? Is there a trick to it during the "sudo make install"  step, where I can have it place the files in the debian folder I'm trying to package instead of the system root?

**Comment by zedtux** [comment]
Well you got the big picture !

But you need to check further because Debian requires files to be here, while Ubuntu requires the same files to be somewhere else.

Basically you start by packaging all yourself, get working packages on many versions (for now the projet is not going forward due to a lack of time on my side), and then you could submit the kernel module (after having learn a lot more about packaging, and after the Douane's DKMS being well tested and stabilised) to the main Debian (so that it will go to Ubuntu too).

To summarise 2 steps :

1/ Package everything by hand (or use services like Travis CI or other stuff dedicated to packaging)
2/ When the soft has gain popularity go with Debian gurus to include it natively

I will check the tutorials I could fine, but anyway, you're on the right path, Debian first, and later Ubuntu (if you want, it's not mandatory, someone else could do it for Ubuntu 😉).

Then I can add you name to [the douaneapp.com](http://douaneapp.com) website 😋

**Comment by thebunnyrules** [comment]
It's a really nice project and it would be my pleasure to help but I'm going to have get much more knowledgeable about packaging before I can help you with it. 

I'll play around with it. I'm still not clear on how I can isolate the built files after I run the make st. For example, say I build douane-daemon.  I do 
`make`
and that build the binaries and then I do 
`make install`
which install the binaries in /. Say I 'm packaging it and I want to files to go into ~/douane-daemon-x.xx-amd64 instead of /, what do I do? There has to be an easier to recreate the folder structure outside of doing make install and then looking for the files installed on my machine. Is there no variation to the _sudo make install_ command that I can use to install the files into the alternate folder? 

I thought of doing chroot but that seems like overkill.

**Comment by cryptoatropine** [comment]
ya i was going to pick this up, a while back, but my horrible net connection and time constraints have put a hamper on it.  forget about check install [in the below link] just skip down to the basic tools for packing ... then follow the link for Debian packages.. linda is no longer a thing ... that is old .... so a good starting point ... but take it as old and maybe in need of updates.  

https://www.linux.com/news/make-your-own-packages-debian-based-systems

**Comment by cryptoatropine** [comment]
i have done done packing before, so i maybe able to help with the learning process as well, 

but i don't always have a net connection, so can sometimes take a long time to respond .. 
That said i would like help in the process as much as i feasibly can.. 

i did make these all into debs a while back, and all of them were easy except the dkms ... never got it to work on more than the system it was built on ...  however that was many months back..

**Comment by thebunnyrules** [comment]
@crytoatrophine, thanks so much for the link. I haven't read through it all but it looks really helpful even if it's old. Thanks for the much appreciated offer of help and I'll definetly take you up once I have any questions. 

I think you may have hit a wall with the dkms because of Secure Boot. As UEFI has become ubiquitous, most PCs come with it enabled by default since the early to mid 2010s.  Hardware with SB enabled will not run or let you install unsigned modules or kernels. I put up a tutorial on signing modules for individual users but it wouldn't work outside the machine where the custom keys are enrolled. Canonical or Debian would have to sign the module themselves to have them run out of the box as their keys are already enrolled at the factory.

**Comment by zedtux** [comment]
Thank you @cryptoatropine for joining here.

You can automate everything using a CI (Like Travis CI for instance) which would checkout the git repositories, run a script doing all the packaging thing.
In this case, you internet connection isn't important 😃

**Comment by zedtux** [comment]
This repo has moved to GitLab and this issue has been imported. You can access it [here](https://gitlab.com/douaneapp/Douane/issues/95).
